### ISSUE TITLE
Clarify CSRF requirement 50.4.1

### ISSUE THREAD

**Issue by tghosth** [OWASP/ASVS#2481]
Current text:

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **50.4.1** | [MODIFIED, MOVED FROM 4.2.2, MERGED FROM 13.2.3] Verify that the application defends against Cross-Site Request Forgery (CSRF) attacks to protect authenticated or sensitive public functionality, using the development framework's built-in anti-CSRF functionality or CSRF tokens, along with additional defense-in-depth measures. | ✓ | ✓ | ✓ | 352 |

We also have:

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **50.4.3** | [ADDED, SPLIT FROM 14.5.3] Verify that the Origin header field is validated against a defined list of allowed origins to match the desired Cross-Origin Resource Sharing (CORS) policy. | ✓ | ✓ | ✓ | 346 |

Points to change:

* Merge origin validation from 50.4.3 into 50.4.1, as discussed here: https://github.com/OWASP/ASVS/issues/876#issuecomment-2550873677
* Mention triggering a CORS preflight

Potential wording:

_[MODIFIED, MOVED FROM 4.2.2, MERGED FROM 13.2.3] Verify that the application defends against Cross-Site Request Forgery (CSRF) attacks to protect authenticated or sensitive public functionality. This should use the development framework's built-in anti-CSRF functionality or CSRF tokens, along with additional defense-in-depth measures such as validating the origin header. Where cross-origin requests are used, this should always trigger a CORS preflight request first._

**Comment by jmanico** [comment]
> [MODIFIED, MOVED FROM 4.2.2, MERGED FROM 13.2.3] Verify that the application defends against Cross-Site Request Forgery (CSRF) attacks to protect authenticated or sensitive public functionality. This should use the development framework's built-in anti-CSRF functionality or CSRF tokens, along with additional defense-in-depth measures such as validating the origin header. Where cross-origin requests are used, this should always trigger a CORS preflight request first.

Notes:

* Simple cross-origin javscript based requests (XHR, fetch) do not trigger pre-flight.
* Cross origin requests from non-JavaScript sources (like HTML forms) do not trigger pre-flight or SOP.

**Comment by randomstuff** [comment]
Yes, "Where cross-origin requests are used, this should always trigger a CORS preflight request first." does not feel right as a consequence.

**Comment by tghosth** [comment]
So I would clarify to:

"_[MODIFIED, MOVED FROM 4.2.2, MERGED FROM 13.2.3] Verify that the application defends against Cross-Site Request Forgery (CSRF) attacks to protect authenticated or sensitive public functionality. This should use the development framework's built-in anti-CSRF functionality or CSRF tokens, along with additional defense-in-depth measures such as validating the origin header. Where cross-origin requests are used **for this functionality**, this should always trigger a CORS preflight request first._"

Can we think of a situation where a cross-origin request for "authenticated or sensitive public functionality" should not trigger a pre-flight?

**Comment by elarlang** [comment]
I don't like the requirement nor the proposal. It takes some time to list all the issues. That's why it is assigned to me with status "awaiting proposal".

**Comment by elarlang** [comment]
So, some ideas to take into account.

Problems:
* The terminology (CSRF) is outdated
  * I think at the time Cross-Site Request Forgery was invented, the term Site was not defined
    * The Site is effective top level domain + 1 (need to take into account the public suffix)
  * Nowadays the term Cross-Site is most of the time used incorrectly
    * Mostly the request forgery happens from the Same-Site scope due to he cookie SameSite attribute setup, but it is called Cross-Site Request Forgery
    * If the request forgery happens due to CORS misconfiguration, the actual problem is Cross-Origin Request Forgery, because the boundary for CORS is Origin (not Site)
* Safe HTTP methods vs state changes - request forgery, or also "use HTTP safe methods" are often limited to "state changing" functionality, but also in the scope should be
  * Resource-demanding functionality, such as data export, serving a large file
  * Authentication form
* Addressed functionality: we need to take into account
  * authenticated and not authenticated functionality
  * state handling with cookies
  * state handling with Authorization header (adding it with JavaScript)
  * state handling with Authorization header (adding by the browser automatically - basic auth, NTLM, etc)

Proposals (to take into account):
* Terminology - we call it abstractly "request forgery"
  * or if to work hand-in-hand with Server-Side Request Forgery then it can also be called Client-Side Request Forgery (ironically, the acronym is CSRF)
* Define the goal to achieve: for executing state-changing or resource demanding functionality (we may ask documentation requirement for that), application must be sure the request was initiated by the application itself or by a trusted party.
* Special highlight for "downgrading" CORS-preflight request downgrade to CORS-safelisted request - need to check Origin and Content-Type request headers.

v4.0.3 Requirements to cover:
* V4.2.2 Verify that the application or framework enforces a strong anti-CSRF mechanism to protect authenticated functionality, and effective anti-automation or anti-CSRF protects unauthenticated functionality.
* V13.2.3 Verify that RESTful web services that utilize cookies are protected from Cross-Site Request Forgery via the use of at least one or more of the following: double submit cookie pattern, CSRF nonces, or Origin request header checks.
* V13.2.5 Verify that REST services explicitly check the incoming Content-Type to be the expected one, such as application/xml or application/json.

Additionally, v5.0.0 requirements to cover:
* V50.4.1 [MODIFIED, MOVED FROM 4.2.2, MERGED FROM 13.2.3] Verify that the application defends against Cross-Site Request Forgery (CSRF) attacks to protect authenticated or sensitive public functionality, using the development framework's built-in anti-CSRF functionality or CSRF tokens, along with additional defense-in-depth measures.
* V50.4.3 [ADDED, SPLIT FROM 14.5.3] Verify that the Origin header field is validated against a defined list of allowed origins to match the desired Cross-Origin Resource Sharing (CORS) policy.
* V13.2.5 Verify that REST services explicitly check the incoming Content-Type to be the expected one, such as application/xml or application/json.

**Comment by randomstuff** [comment]
> The terminology (CSRF) is outdated 

I agree. I don't like this term very much (and I don't like "XSS" much either for the same kind of reasons). However, this is the accepted wording. It is difficult to depart from it and mint your own wording when the whole community is using this "CSRF".

FWIW, I think a better wording would be "cross origin request forgery" but it is not used at all anywhere.

> then it can also be called Client-Side Request Forgery 

I kind-of like that :smile:.

I'm wondering if this is clear enough though. There might be other kind of client-side request forgery (for example, client-side cross-protocol confusion attacks, [ALPACA](https://alpaca-attack.com/ALPACA.pdf)).

**Comment by elarlang** [comment]
Offtopic:
> and I don't like "XSS" much either for the same kind of reasons

Probably you can find also from here in issue discussions how much I don't like the XSS. The XSS even worse than CSRF. CSRF is "half ok", because the "request forgery" part still explains what it is. But XSS is also used in "Same-Origin HTML injection" situation, which means not a single word from cross-site scripting is actually true, and this is insane nonsense.

---

Ontopic:
> However, this is the accepted wording. It is difficult to depart from it and mint your own wording when the whole community is using this "CSRF".

I think "this is the used term", not that much maybe "the accepted term" anymore. At the end, the term Cross-Site Request Forgery is incorrect and there is reason to use it. The same with XSS - we have it mentioned only in one requirement in the context of impact. The "Cross-Site" is used also in "XSSI" name.

> FWIW, I think a better wording would be "cross origin request forgery" but it is not used at all anywhere.

I would not limit it to cross-origin either. It can be also same-origin - for example, an attacker can use only HTML injection (classical outcome from sanitized HTML, you can still use `<img src="/local/path">`), but with that can trigger all functionality (if cookies are used for state handling) using GET method - for example calling resource demanding export, call logout etc.

**Comment by tghosth** [comment]
@elarlang I see your point but can you make a concrete wording suggestion please.

**Comment by jmanico** [comment]
Agreed XSS is not a good term. All three of the words can me wrong. 

Many folks use “web UI content injection”, “content injection” or similar as a more accurate term.

**Comment by jmanico** [comment]
Addressing proposals from @elarlang :

1) Terminology - we call it abstractly "request forgery"

**Cross Site Request Forgery**,  **Session Riding** CSRF and XSRF are the common terms and acronyms. If we stray beyond these, can we please cite them along side the new term? Maybe session riding is more accurate that cross site request forgery or request forgery? But overall I do not think this is a big problem. Most of the industry recognizes these legacy terms. The generic term "request forgery" is _**too generic IMO**_, it can incorrectly refer to SSRF or Clickjacking.

2) Define the goal to achieve: for executing state-changing or resource demanding functionality (we may ask documentation requirement for that), application must be sure the request was initiated by the application itself or by a trusted party.

I like the direction of this. But I think it's more than state changing functionality. I think it's any functionality that would cause damage if the request was forged, regardless of the verb or state changing effect. 

3) Special highlight for "downgrading" CORS-preflight request to CORS-safelisted request - need to check Origin and Content-Type request headers.

I think this is a really good idea to add in. Thumbs up on this one.

**Comment by elarlang** [comment]
Previously I proposed to use client-side request forgery to be "paired" with SSRF, but one of them is making the request and another is accepting the request:

  * SSRF - The defense against SSRF is to not allow the application server-side component to _make_ the request to the location based on untrusted input
  * CSRF - The defense against CSRF is to not allow the application server-side component to _accept_ a request from an untrusted party

The "Client" is also vague, and in the situation of SSRF to happen, the vulnerable application itself is a client to make the request. So maybe browser-based request forgery is a more correct or self-explaining term to use.

---

Part 1 - the definition:
> Verify that the application validates the request to call out state-changing or sensitive functionality originating from the functionality provided by the application itself or by a trusted party, to defend against browser-based request forgery attacks (previously known as cross-site request forgery (CSRF)).

---

Part 2 - the clarification the functionality to defend

> Sensitive functionality includes accepting form posts for authenticated and non-authenticated users (such as authentication), calling out resource-demanding functionality (such as data export), and (todo: what else to list here?)

---

Part 3 - CORS spotlight

> For requests that expect to trigger a CORS-preflight request, request header fields such as 'Origin' and 'Content-Type' must be validated to ensure that it is not possible to call the functionality with CORS-safelisted request.

---

Part 4 - terminology note

Too long to add into the requirement, probably part of the chapter text:
> Note that the attack vector is known as cross-site request forgery (CSRF), but the definition of site (effective Top Level Domain + 1) makes the term CSRF outdated and incorrect. The scope limitation to cross-site is not valid, as the attack may come from 'same-site but cross-origin' or from 'same-site and same-origin' scope.

**Comment by elarlang** [comment]
We also have a requirement for safe-methods:
> V13.6.2 [MODIFIED, MOVED FROM 13.2.1] Verify that HTTP requests using the HEAD, OPTIONS, TRACE or GET verb do not modify any backend data structure or perform any state-changing actions. These requests are safe methods and should therefore not have any side effects.	

The main goal is the same, it is a kind of pre-condition to have a defense against browser-based request forgery attacks and it does not have its own separate risk.

The choice here is, to merge everything into one monster requirement, or to have some more specific separate requirements to address separate related and underlying problems, such as using safe methods for state changes or CORS-preflight downgrade to CORS-safelisted.

**Comment by tghosth** [comment]
So what is the rough wording suggestion please here @elarlang?

**Comment by elarlang** [comment]
> So what is the rough wording suggestion please here @elarlang?

Rough wording is concat(part1, part2, part3, part-x) - written as separate parts to have more precise feedback and finetuning for different parts.

**Comment by tghosth** [comment]
Regarding,


> the term "sensitive functionality" - there should be something better, as it may get mixed or confusion with data classification.

Maybe we need to focus on state changing or resource intensive functionality. Not sure I know what sensitive functionality means in this that isn't included by one of those other definitions.

> part 2 - what else must be listed to "to be protected against "csrf" list?

Not convinced there is anything else.

I would propose the following text:



"_Verify that requests to state-changing or resource intensive functionality are checked to ensure they originate from the application itself, to defend against browser-based request forgery attacks (commonly known as cross-site request forgery (CSRF)). Additionally, this functionality must not be invoked using safe HTTP methods such as HEAD, OPTIONS, or GET. This precaution will ensure that cross-origin requests to this functionality will trigger a preflight request._"


Comments on the proposal:

* In this use case, the "trusted party" is considered part of the application.
* _Part 2 - the clarification the functionality to defend_ - I think this part is now redundant with the updated wording.
*  _(e.g. from a navigation request)_ - I think this is redundant or should be moved to the section text.
* _Part 4 - CORS spotlight (merge 13.2.5)_ - Based on what it says [at this link](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#simple_requests), we are enforcing preflights through our HTTP verb requirement.
* _Part 5 - terminology note_ - I have included a proposal below to rework the section text.
Section text:

### V50.4 Browser Origin Separation

When accepting a request on the server side, we need to be sure it is initiated by the application itself and not in some way spoofed by a malicious party. Key security mechanisms here include browser security policies like Same Origin Policy for JavaScript and also SameSite logic for cookies.

The most common and well known attack vector is known as cross-site request forgery (CSRF), but note that the definition of site (effective Top Level Domain + 1) makes the term CSRF outdated and incorrect. The scope limitation to cross-site is not valid, as the attack may come from 'same-site but cross-origin' or from 'same-site and same-origin' scope.

**Comment by elarlang** [comment]
> In this use case, the "trusted party" is considered part of the application.

As we included CORS, it may be some other trusted party that uses the API

> Part 2 - the clarification the functionality to defend - I think this part is now redundant with the updated wording.

The proposal lost spotlight for the non-authenticated actions, such as authentication.

> Part 4 - CORS spotlight (merge 13.2.5) - Based on what it says [at this link](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#simple_requests), we are enforcing preflights through our HTTP verb requirement.

If you don't validate the previously/my proposed part 4, you can bypass this new/your proposed part 4.

**Comment by tghosth** [comment]
I think the following requirement covers parts 1, 2 and 4 above:

"_Verify that requests to state-changing, security sensitive, or resource intensive functionality includes a mechanism to prevent browser-based request forgery attacks, commonly known as cross-site request forgery (CSRF). Ideally the functionality should only accept content-types which would trigger a CORS preflight and enforce an allow list of 'Origin' headers. Alternatively, another CSRF prevention mechanism may be used._"

It covers the "write" case here where the "read" case is covered by [50.3.6](https://github.com/OWASP/ASVS/blob/master/5.0/en/0x50-V50-Web-Frontend-Security.md#v503-browser-security-mechanism-headers).

I think the following requirement covers parts 1, 2 and 4 above:

Verify that requests to state-changing, security sensitive, or resource intensive functionality use appropriate HTTP methods including POST, PATCH, DELETE or PUT. 

Part 5 will be covered in chapter text but we have a note that we wanted to mention authentication as a specific "write" case there.

**Comment by jmanico** [comment]
The only thing that tweaks me here is that SOP in terms of CSRF protection only matters for API’s and similar, standard same-site websites do not benefit from SOP for CSRF protection.

The “alternative” defense suggested in the last sentence is not alternative, its fundamental for standard websites and this requirement lends me to believe that SOP and preflight is enough when it often is not for CSRF defense.

**Comment by ryarmst** [comment]
I would like to add my opinion that I think it is a good opportunity to move away from Cross-Site Request Forgery as a term. That said, our field would not be unique in clinging to incorrect terminology. "Heartburn" for example, has nothing to do with the heart...

> Maybe we need to focus on state changing or resource intensive functionality. Not sure I know what sensitive functionality means in this that isn't included by one of those other definitions.

I would consider XS Leaks and (See #2409), Spectre-like attacks (also #2409), and attacks like BREACH (#2203) to be relevant to the discussion. Here is a relevant [comment](https://github.com/OWASP/ASVS/issues/2203#issuecomment-2464361620) from the closing of #2203:

> We also think that other cross origin and anti-automation controls would hopefully help mitigate this.

I think therefore it is worth consider these classes of issues when constructing requirements to restrict processing of cross-site or cross-origin requests because there is functionality beyond resource intensive or state changing that could be impacted. Unfortunately, that also complicates the considerations. Thoughts?

One more thing I would like to add to the discussion (sorry): [Fetch Metadata ](https://developer.mozilla.org/en-US/docs/Glossary/Fetch_metadata_request_header) is now widely supported and provides much more contextual information about the nature of requests (compared to just the Origin header).

**Comment by elarlang** [comment]
Let's focus for this issue only to handle the related requirements (50.4.1, 13.2.5, 13.6.2).

If there are related requirements, let's have the wording first and then we can see, where those fits.

I'm aware of `Sec-Fetch-*` header fields, those are in use also in requirement 50.6.1. I feel it is more #2049 area to discuss.

---

On topic: As "the one monster csrf requirement" does not work out, I think we should move away from this direction and keep current separate requirements with finetuned focuses. I'll come back to this with a separate comment/proposal..

**Comment by tghosth** [comment]
@elarlang to prepare an updated proposal

**Comment by tghosth** [comment]
#thiscommentshouldhavebeenablogpost

**Comment by tghosth** [comment]
Please can you add your proposals to the tl;dr together with tags to help understand which current requirements they are modifying.

**Comment by elarlang** [comment]
In the current situation, the long post was (forced to be) written to explain and address all the situations, different test cases, and expected developments together and point out, how they don't overlap.

I recommend taking the time and analyze it. Understanding it is a pre-condition to make any decision based on the provided information. So I don't think those 3 proposed requirements should be taken out of context and then start answering the questions out of context that are actually already written in the post.

**Comment by tghosth** [comment]
Ok so as I understand it, your proposed requirements are:

Part 4:

V50.4.3 - [MODIFIED, MOVED FROM 13.2.5, SPLIT FROM 14.5.3] Verify that if the application uses and relies on CORS-preflight mechanism, it is not possible to call the functionality with a CORS-safelisted request. It may require validating the 'Origin' and 'Content-Type' request headers or using an extra header field in the communication that is not a CORS-safelisted request header field.

Part 3:

V13.6.2 - [MODIFIED, MOVED FROM 13.2.1] Verify that state-changing or resource-demanding functionality is not executed for a navigation request or is called an incorrect destination (e.g. export data functionality is called as an image from the browser). Not accepting safe HTTP methods such as HEAD, OPTIONS, or GET and Sec-Fetch-* request headers can be used for validation. It requires special attention if the application does not distinguish URL parameters and message body parameters.

Part 1:

50.4.1 - [MODIFIED, MOVED FROM 4.2.2, MERGED FROM 13.2.3] Verify that requests to state-changing or resource intensive functionality are checked to ensure they originate from the application itself. For example, by using and validating tokens or extra HTTP headers that are not CORS-safelisted request-headers. This is to defend against browser-based request forgery attacks, commonly known as cross-site request forgery (CSRF).

Correct? If so, I can then go and analyse your argumentation whilst clearly understanding your end goal from your argumentation.

**Comment by elarlang** [comment]
Relations to current requirements that will be covered with new proposal, were listed in my post into each part. So I think the question was answered already.

How correct tags are pre-condition to analyze the concept "one monster requirement" vs "3 separate requirements"  is not understandable for me. What is the correct tag also depends on the requirement. I listed at the moment current v5.0.be requirements, if we want to have correct tags, we need to analyze the content of a v4.0.3 requirement compared to new proposed requirement. At this stage, I don't see the reason to put time into it. Just for an example, 13.2.3 from v4.0.3 is basically covered by 2 separate requirements.

Overall, requirements to cover was provided in https://github.com/OWASP/ASVS/issues/2481#issuecomment-2558410664 and 13.2.6 was linked later in https://github.com/OWASP/ASVS/issues/2481#issuecomment-2567821725

**Comment by tghosth** [comment]
**Summary of discussion** 

----------------------------------------------------

**These are ok:**

Part 4:

"_V50.4.3 - [MODIFIED, MOVED FROM 13.2.5, SPLIT FROM 14.5.3] Verify that, if the application relies on the CORS preflight mechanism to prevent disallowed cross-origin use of sensitive functionality, it is not possible to call the functionality with a CORS-safelisted request. This may require checking the values of the 'Origin' and 'Content-Type' request headers or using an extra header field that is not CORS-safelisted._"

Add a note to chapter text that this applies to apps that are designed to be accessed cross origin but also those which are not but instead preflight is used as a security mechanism.


Part 1:

"_50.4.1 - [MODIFIED, MOVED FROM 4.2.2, MERGED FROM 13.2.3] Verify that CORS-safelisted requests to sensitive functionality are checked to ensure that they originate from the application itself. This may be done by using and validating anti-forgery tokens or requiring extra HTTP headers that are not CORS-safelisted request-headers. This is to defend against browser-based request forgery attacks, commonly known as cross-site request forgery (CSRF)._"

Add anti-forgery tokens to glossary?

-------------------------------

**This needs more thought:**


Part 3:

V13.6.2 - [MODIFIED, MOVED FROM 13.2.1] Verify that sensitive functionality is not executed for a navigation request or is called an incorrect destination (e.g. export data functionality is called as an image from the browser). Not accepting safe HTTP methods such as HEAD, OPTIONS, or GET and Sec-Fetch-* request headers can be used for validation. It requires special attention if the application does not distinguish URL parameters and message body parameters.

**Comment by tghosth** [comment]
Looking again at Part 3:

This is the current state:

"_V13.6.2 - [MODIFIED, MOVED FROM 13.2.1] Verify that HTTP requests using the HEAD, OPTIONS, TRACE or GET verb do not modify any backend data structure or perform any state-changing actions. These requests are safe methods and should therefore not have any side effects._"

I previously said ([here](https://github.com/OWASP/ASVS/issues/2492#issuecomment-2567893752)) that I thought this should be merged into a separate CSRF requirement. I am prepared to accept that it needs to stay separate and also include the subtleties of the version [proposed above](https://github.com/OWASP/ASVS/issues/2481#issuecomment-2609670196).

I therefore propose the following @elarlang :

"_V13.6.2 - [MODIFIED, MOVED FROM 13.2.1] Verify that calls to sensitive functionality use appropriate HTTP methods such as POST, PUT, PATCH or DELETE, with methods defined by the HTTP specification as "safe" such as HEAD, OPTIONS, or GET not being used. Alternatively, strict validation of the Sec-Fetch-\* can be used to ensure that the request did not originate from an inappropriate cross-origin call, a navigation request, or a resource load (such as an image source). This is particularly important if the application does not distinguish between URL parameters and message body parameters_."

What do you think?

**Comment by elarlang** [comment]
Proposed middle-part:
> Alternatively, strict validation of the Sec-Fetch-* can be used to ensure that the request did not originate from an inappropriate cross-origin call, a navigation request, or a resource load (such as an image source).

Attempt to turn it into positive requirement:
> Alternatively, strict validation of the Sec-Fetch-* can be used to verify that the request to sensitive functionality was initiated the way the application expects it (such as not accepting cross-origin request, navigation request, or executing functionality as called by image source).

If to keep the first option, "_such as an image source_" needs to be updated to say that "_executing functionality being loaded as image resource is not accepted if not intended_".

**Comment by tghosth** [comment]
> or executing functionality as called by image source

I don't understand why this level of detail is necessary? We have already said this is about calling sensitive functionality. When would we ever expect this from a resource load?

**Comment by elarlang** [comment]
Images and dynamically creating images can be called as an image resource and it is expected functionality.

**Comment by tghosth** [comment]
"_V13.6.2 - [MODIFIED, MOVED FROM 13.2.1] Verify that calls to sensitive functionality use appropriate HTTP methods such as POST, PUT, PATCH or DELETE, and not methods defined by the HTTP specification as "safe" such as HEAD, OPTIONS, or GET. Alternatively, strict validation of the Sec-Fetch-* can be used to ensure that the request did not originate from an inappropriate cross-origin call, a navigation request, or a resource load (such as an image source) **where this is not expected**. This is particularly important if the application does not distinguish between URL parameters and message body parameters._"

**Comment by elarlang** [comment]
One more finetuning:
> Verify that calls to sensitive functionality use appropriate HTTP methods such as POST, PUT, PATCH or DELETE, and not methods defined by the HTTP specification as "safe" such as HEAD, OPTIONS, or GET. Alternatively, strict validation of the Sec-Fetch-* **request header fields** can be used to ensure that the request did not originate from an inappropriate cross-origin call, a navigation request, or a resource load (such as an image source) where this is not expected. This is particularly important if the application does not distinguish between URL parameters and message body parameters

**Comment by tghosth** [comment]
Ok so I think we are there, I think all these requirements now belong in V50.4 so I would propose as follow:


## V50.4 Browser Origin Separation

When accepting a request on the server side, we need to be sure it is initiated by the application itself or by a trusted party and has not been forged by an attacker. The keywords here are browser security policies like Same Origin Policy for JavaScript and also SameSite logic for cookies. Another common protection is the CORS preflight mechanism. This mechanism will be critical for endpoints designed to be called from a different origin, but it can also be a useful request forgery prevention mechanism for endpoints which are not designed to be called from a different origin.

The category should contain requirements with ideas:

* Verify that the request was initiated by a trusted party (CSRF, CORS misconfiguration)
* Verify that the response is readable only for trusted parties (CORS misconfiguration)

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **50.4.1** | [MODIFIED, MOVED FROM 4.2.2, MERGED FROM 13.2.3] Verify that CORS-safelisted requests to sensitive functionality are checked to ensure that they originate from the application itself. This may be done by using and validating anti-forgery tokens or requiring extra HTTP headers that are not CORS-safelisted request-headers. This is to defend against browser-based request forgery attacks, commonly known as cross-site request forgery (CSRF). | ✓ | ✓ | ✓ | 352 |
| **50.4.3** | [MODIFIED, MOVED FROM 13.2.5, SPLIT FROM 14.5.3] Verify that, if the application relies on the CORS preflight mechanism to prevent disallowed cross-origin use of sensitive functionality, it is not possible to call the functionality with a CORS-safelisted request. This may require checking the values of the 'Origin' and 'Content-Type' request headers or using an extra header field that is not CORS-safelisted. | ✓ | ✓ | ✓ | 346 |
| **50.4.4** | [MODIFIED, MOVED FROM 13.2.1] Verify that calls to sensitive functionality use appropriate HTTP methods such as POST, PUT, PATCH or DELETE, and not methods defined by the HTTP specification as "safe" such as HEAD, OPTIONS, or GET. Alternatively, strict validation of the Sec-Fetch-* request header fields can be used to ensure that the request did not originate from an inappropriate cross-origin call, a navigation request, or a resource load (such as an image source) where this is not expected. This is particularly important if the application does not distinguish between URL parameters and message body parameters. | ✓ | ✓ | ✓ | 650 |


---

# V13

## V13.2 Web Services

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **13.2.1** | [MOVED TO 50.4.4] | | | | |
| **13.2.3** | [DELETED, MERGED TO 50.4.1] | | | | |
| **13.2.5** | [MOVED TO 50.4.1] | | | | |

## V13.6 HTTP Request Header Validation

Delete 13.6.2 and move the two reqs after it up by one number

---

# Appendix A: Glossary

* **Anti-forgery token** - A mechanism by which one or more tokens are passed in a request and validated by the application server to ensure that the request has come from an expected endpoint.

**Comment by tghosth** [comment]
Let me know if you are good with this @elarlang and I will open a PR and make notes on the other relevant issues

**Comment by elarlang** [comment]
The chapter text should contain the explanation what kind of functionality must be protected and it is valid for authenticated and non-authenticated users.

I need to rethink the modification tags, but you can start with the PR.
